client/state/utils.js

```diff
@@ -25,6 +25,22 @@ export function isValidStateWithSchema( state, schema, checkForCycles = false, b
 }
 
 /**
+ * Given an action object or thunk, returns an updated object or thunk which
+ * will include additional meta in the action (as provided) when dispatched.
+ *
+ * @param  {(Function|Object)} action Action object or thunk
+ * @param  {Object}            meta   Additional meta to include in action
+ * @return {(Function|Object)}        Augmented action object or thunk
+ */
+export function dispatchWithMeta( action, meta ) {
+	if ( 'function' !== typeof action ) {
+		return { ...action, meta };
+	}
+
+	return ( dispatch ) => action( ( thunkAction ) => dispatch( { ...thunkAction, meta } ) );
+}
```

Comment:
The way we are setting arbitrary values to the `meta` property is a bit unsettling to me here. In my head I feel like what goes into meta should both be isolated by some upper-level name into categories of things that will handle it (most likely different middleware pieces) and also probably stored as arrays, since theoretically we could keep on adding more meta.

When I see how this is used, for example, I don't see a generic system for adding meta into an action, but rather a way of tacking specific data onto an existing action through the thunk barrier. Perhaps it might be described as `extendAction`.